core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java

```diff
@@ -369,6 +379,33 @@ private void modifyColumnSchemaForSortColumn(ColumnSchema columnSchema, boolean
     return tableBlockIndexUniqueIdentifiers;
   }
 
+  /**
+   * This case is added for a case where, there are two applications running, and in one application
+   * operations happened like SI rebuild, update or delete case, then the cache should be updated as
+   * well. The cache updation happens for same application, but other application may fail to query
+   * or may give wrong result. Since we overwrite the segment file in these scenarios, check the
+   * timestamp, and if modified, clear from the cache.
+   */
+  private void clearSegmentMapIfSegmentUpdated(String latestSegmentFilePath, Segment segment) {
+    SegmentBlockIndexInfo segmentBlockIndexInfo = segmentMap.get(segment.getSegmentNo());
```

Comment:
For a clean approach, 

Need to use `SegmentRefreshInfo`, for transactional table , only update scenario will use this. can use for normal scenario also. For nontransactional carbon table. Already it is used for this non-update scenario. Refer `LatestFilesReadCommittedScope#takeCarbonIndexFileSnapShot` to see how it is filled.

Reply:
But that case, we will be doing list files, get all the files inside segment, timestamp comparison, i  dont think this will be feasible, here, just i can go for segment file, one file timestamp checking would be feasible and in future also, any segment file related updation will be taken care.

Reply:
That changes done way before segment file, now no need to list index files. you can use current changes but fill that timestamp in `SegmentRefreshInfo`. It will take care of cache clean up part

Reply:
this was done for non transactional table, in transactional case, we have scenarios where we have millions of files inside a segment, so i think this approach will be better with respect to performance, and in future all operations will be based on segment file, so i think it will be feasible. what u think?

Reply:
As discussed, we can use SegmentRefreshInfo interface in this scneario (no need to list index file, can get info from segment file)

Reply:
please refer #3686 i have created new PR

Reply:
It was WIP, would have handled in the same PR

Reply:
its k, no issues